### ISSUE TITLE
[RFC] Add support for missing OPAM fields

### ISSUE THREAD

**Issue by Leonidas-from-XIV** [ocaml/dune#7059]
## Desired Behavior

Dune users at the moment have an convenient escape-hatch to add various OPAM fields to their generated `<package>.opam` files via `<package>.opam.template` files. This is nice as OPAM supports more fields than dune, but it would be nice if we could make the gap between Dune and OPAM smaller by supporting to generate all/virtually all fields that OPAM can support at the moment (some fields don't make sense like `build` or `remove`).

Unfortunately, the Dune documentation doesn't state which fields are supported, which this RFC also suggests to amend.

The [OPAM manual](https://opam.ocaml.org/doc/Manual.html#opam) has a good list of possibly allowed keys to have support for. It is of course possible to add `x-extension` keys to OPAM files, but these can safely remain unimplemented for now - the `.template` files should be good enough for this usecase.

## Example

Being allowed to specify 

```sexp
(tags "org:erratique" "command-line")
(pin-depends
  (name url)
  (name' url')
  (name'' url'')
```

and getting the apprpriate translation of it in the `OPAM` file(s).

**Comment by emillon** [comment]
Some remarks:
- tags is supported (I thought it required a specific version or something but I checked the source and it doesn't seem like it)
- pin-depends is the most important one to support
- I don't think we should blindly add all supported fields that opam supports. there are many of them that are actually not being used in the wild and have unclear semantics or even syntax (for each supported field we need to have a way to express that as dune-lang - it's an unnecessary burden for most fields)
- for some fields, we support the field but with a less general syntax. `depends` in particular does not support all kinds of formulas that opam does. going for the full expressivity of filtered formulas means that some operations that dune does like "add a lower bound when generating the opam file" can become harder to implement. this is something we need to clarify.

So in a nutshell, I think that this is a good idea but as written this is an issue that is difficult to mark as complete because the scope is very large. I suggest that we restrict and clarify the scope to something like "pin-depends and generic formulas with the restriction that constraints are only added to naked dependencies". We can find the useful subset by `opam-grep`ing for `.opam.template` files and looking at the most important missing features.

**Comment by rgrinberg** [comment]
The pin depends field has been rejected on purpose to allow for dune to have its own package management one day.

> I don't think we should blindly add all supported fields that opam supports. there are many of them that are actually not being used in the wild and have unclear semantics or even syntax (for each supported field we need to have a way to express that as dune-lang - it's an unnecessary burden for most fields)

In general, for the field to be included, it has to be justified beyond "I want to generate an opam file with it". In should be argued why this field makes sense from a package management point of view and why the project file is the correct place to put it.

> depends in particular does not support all kinds of formulas that opam does. going for the full expressivity of filtered formulas means that some operations that dune does like "add a lower bound when generating the opam file" can become harder to implement

Correct. The limitation is intentional as well.

**Comment by Leonidas-from-XIV** [comment]
@emillon That's a good point. Using the cache that `opam-cache` has generated on my machine and a slightly hackish `find -name '*.opam.template' | xargs cat | grep -Eo '^[a-zA-Z-]*:' | sort | uniq -c | sort -bg | tac` invocation, I get this histogram of fields used in `.template` files:

```
    343 build:
    115 available:
     24 license:
     19 tags:
     18 depends:
     14 dev-repo:
     12 depexts:
     12 bug-reports:
     11 substs:
      6 x-ci-accept-failures:
      6 install:
      5 flags:
      5 doc:
      4 post-messages:
      4 conflicts:
      3 homepage:
      2 version:
      2 synopsis:
      2 pin-depends:
      2 name:
      2 messages:
      2 maintainer:
      2 description:
      2 build-env:
      2 authors:
      1 x-foo:
      1 run-test:
      1 opam-version:
      1 libraries:
```

A few observations:

  * Some packages have their own `build` stanzas. It's worth looking into how many of those are necessary and how many of these could be handled by `dune` already.
  * `license` is mostly either people being unaware of the field being supported or unaware that multiple licenses can be specified (I've created #7098 to make it a bit clearer in the documentation). Given we have plural `authors` and `maintainers` maybe having an alias `licenses` could be helpful?
  * `tags`: people like me unaware that this is supported. Only affects 3 packages (AltErgo, CA-Certs, Fiat)
  * `depends`: only `dune` packages. Also odoc-parser which legitimately needs to do this to not create a circular dependency on itself.

[Will continue to update later]

**Comment by rgrinberg** [comment]
>  Some packages have their own build stanzas. I'll look closer into this.

Mostly packages that want to customize their sources for release builds. It would be nice to do this in dune directly.

> depexts: This seems to be a legitimate omission that might be worth implementing

Indeed it would be nice. Note that I omitted depexts because there's nothing that dune can do about satisfying them. In particular, running `brew` or `apt` isn't something that's reasonable to do within dune. We could just add them and say they're just pure metadata for opam, but I would prefer a separation between fields that dune understands and raw metadata.

**Comment by emillon** [comment]
Regarding depexts, I believe it's best practice to only have these on conf- packages, not directly on buildable packages. These conf packages do not need to build with dune, and in the rare reason where a depext is necessary, template files work.

**Comment by Leonidas-from-XIV** [comment]
> Mostly packages that want to customize their sources for release builds. It would be nice to do this in dune directly.

Yes. From what I see there are different categories:

  1. `build` instructions that could be generated by dune already. `camelot.1.7.1` is one of these cases where I don't see any customization.
  2. `build` instructions where some kind of configuration script is run beforehand. `camomile.1.0.2` is one example, AltErgo is another example, but they also leave out some part of the dune command
  3. Some customize the `@runtest` (to either use a different alias or not run on certain systems). I remember having a similar usecase. I think this is worth delving into how we can enable people to solve this in a nicer way.

> We could just add them and say they're just pure metadata for opam, but I would prefer a separation between fields that dune understands and raw metadata.

We already have that to some degree, `authors`, `maintainers`, `tags` are only used to generate OPAM data.

I also looked into `available` and most of the cases it is about limiting availability on the OS/distribution. I think that's also a useful feature for dune, where `dune` could also warn users if they try to build on unsupported platforms.

Let's decide which flags we should add, then I can write proposals for syntax on them and implement them when we have a consensus.

**Comment by rgrinberg** [comment]
> build instructions where some kind of configuration script is run beforehand. camomile.1.0.2 is one example, AltErgo is another example, but they also leave out some part of the dune command

It would be nice to see why the configuration isn't done directly in dune rules.

> Some customize the @runtest (to either use a different alias or not run on certain systems). I remember having a similar usecase. I think this is worth delving into how we can enable people to solve this in a nicer way.

I think there's already `enabled_if` for that?

> I also looked into available and most of the cases it is about limiting availability on the OS/distribution. I think that's also a useful feature for dune, where dune could also warn users if they try to build on unsupported platforms.

I think this can be expressed well enough with existing concepts. For example, we can have `conf-linux` to express that a package is linux only. Moreover, if we ever decide to make this first class, we should express such constraints using dune's language for dependencies. That way, we'll have a natural progression from enabled_if to project level dependencies.

**Comment by rgrinberg** [comment]
> We already have that to some degree, authors, maintainers, tags are only used to generate OPAM data.

Yes, that's true - for now. I would like to make this metadata more useful by generating more than just opam files, making it a part of `dune describe`, and allowing it to be accessed via `%{variables}` eventually.
